lib/refinement/annotated_target.rb

```diff
@@ -0,0 +1,63 @@
+module Refinement
+  # A target, annotated with any changes
+  class AnnotatedTarget
+    # @return [Xcodeproj::Project::AbstactTarget] the target in an Xcode project
+    attr_reader :xcode_target
+    # @return [String,Nil] the reason why the target has changed, or `nil` if it has not changed
+    attr_reader :change_reason
+
+    def initialize(target:, change_reason:, dependencies: [])
+      @xcode_target = target
+      @change_reason = change_reason
+      @dependencies = dependencies
+      dependencies.each do |dependency|
+        dependency.depended_upon_by << self
+      end
+      @depended_upon_by = []
+    end
+
+    # @visibility private
+    def to_s
+      xcode_target.to_s
+    end
+
+    CHANGE_LEVELS = %i[
+      itself
+      at_most_n_away
+      full_transitive
+    ].freeze
+    private_constant :CHANGE_LEVELS
+
+    # @return [Boolean] whether the target has changed, at the given change level
+    # @param level [Symbol,(:at_most_n_away,Integer)] change level, e.g. :itself, :at_most_n_away, :full_transitive
+    def changed?(level:)
+      @changed ||= {}
+      return @changed[level] if @changed.key?(level)
+
+      @changed[level] =
+        case level
+        when :itself
+          change_reason
```

Comment:
seems like we are expecting a boolean result but returning a `String` or `nil` here.

Reply:
actually I am wrong this is about the assignment of `@changed[level]` line but I believe the same still applies.